### PR TITLE
[Graph] Support lookup of incoming edges

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @lawzlo 
This should enable lighter data models.

### CODE REVIEW

**Comment by leogrim** [comment]
Note that serialization doesn't change, so a simple restart will have this working. No need for reingesting.

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraphReader.scala [hunk 1]
```diff
@@ -18,70 +26,92 @@ object Vertex {
...
-  def destination: VertexId = currentDestination getOrElse { throw new UninitializedReaderException(s"$this is not initialized over a valid destination vertex") }
-  def data: EdgeDataReader = edges(component)(destination)
-  def kind: EdgeKind[_ <: EdgeDataReader] = data.kind
+trait SimpleLocalEdgeReader extends LocalEdgeReader {
+  protected def getComponentIterator(): Iterator[(VertexType, VertexType, EdgeType)]
+  protected def getEdgeIterator(): Iterator[(VertexId, VertexId, EdgeDataReader)]
+  private var components: Option[Iterator[(VertexType, VertexType, EdgeType)]] = None
+  private var currentComponent: Option[(VertexType, VertexType, EdgeType)] = None
+  private var edges: Option[Iterator[(VertexId, VertexId, EdgeDataReader)]] = None
+  private var currentEdge: Option[(VertexId, VertexId, EdgeDataReader)] = None
```
Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraphReader.scala:59]: Not sure how much garbage these tuple iterators will generate, but they enable to abstract the iteration logic and make the code much cleaner. If a problem (meh), we can always separate the outgoing and incoming implementations to iterate over either sources or destinations like I used to for outgoing edges.

**Comment by stkem** [comment]
lgtm

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraphReader.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraphReader.scala:59]: yeah, something we'll likely have to revisit at some point, but I',m not too concerned right now

**Pull request merged**